docs/training/apis-howto/api-core-mnist-ug.rst

```diff
@@ -0,0 +1,573 @@
+``train.report_training_metrics()``:
+
+Use ``core_context.train`` to report training and validation metrics.
+
+.. note::
+
+   Refer to line 69 in ``model_def_metrics.py``
+
+.. code::
+
+   # NEW: Report training metrics to Determined master via core_context.
+   # Index by (batch_idx + 1) * (epoch-1) * len(train_loader) to continuously plot loss on one graph for consecutive epochs.
+   core_context.train.report_training_metrics(
+       steps_completed=(batch_idx + 1) + (epoch - 1) * len(train_loader),
+       metrics={"train_loss": loss.item()},
+
+.. note::
+
+   Refer to line 101 in ``model_def_metrics.py``
+
+.. code::
+
+   # NEW: Report validation metrics to Determined master via core_context.
+   core_context.train.report_validation_metrics(
+       steps_completed=steps_completed,
+       metrics={"test_loss": test_loss},
+   )
+
+.. tip::
+
+   Avoiding a Duplicate Key Error
+
+   It is best to stitch metrics together in one continuous graph as your experiment progresses. To
+   accomplish this, calculate the index used for ``steps_completed`` appropriately. The appropriate
+   calculation avoids re-writing metrics with the same index each time a new epoch begins and avoids
+   a duplicate key error.
+
+Step 2.3: Modify the Test Method
+================================
+
+Modify the ``test()`` function header to include ``args`` and other elements you’ll need during the
+evaluation loop. In addition, pass the newly created ``core_context`` into both ``train()`` and
+``test()``.
+
+.. code::
+
+   # NEW: Pass core_context into train() and test().
+   train(args, model, device, train_loader, optimizer, epoch, core_context)
+
+   # NEW: Pass args, test_loader, epoch, and steps_completed into test().
+   test(args, model, device, test_loader, epoch, core_context, steps_completed=steps_completed)
+
+Create a ``steps_completed`` variable to plot metrics on a graph in the WebUI:
+
+.. note::
+
+   Refer to line 198 in ``model_def_metrics.py``
+
+.. code::
+
+   # NEW: Calculate steps_completed for plotting test metrics.
+   steps_completed = epoch * len(train_loader)
+
+Step 2.4: Run the Experiment
+============================
+
+Run the following command to run the experiment:
+
+Finally, run the following command to run the experiment:
+
+.. code::
+
+   det e create metrics.yaml .
+
+Open the Determined WebUI again and navigate to the **Overview** tab.
+
+The WebUI now displays metrics. In this step, you learned how to add a few new lines of code in
+order to report training and validation metrics to the Determined master. Next, we’ll modify our
+script to report checkpoints.
+
+***********************
+ Step 3: Checkpointing
+***********************
+
+Checkpointing periodically during training and reporting the checkpoints to the master gives us the
+ability to stop and restart training. In this section, we’ll modify our script for the purpose of
+checkpointing.
+
+..
+   note::Two Types of Continuations (Stopping and Restarting)
```

Comment:
sphinx is reading this as a comment, not as a `note` directive.

Reply:
resolved